api/src/main/java/org/apache/iceberg/io/RangeReadable.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * {@code RangeReadable} is an interface that allows for implementations
+ * of {@link InputFile} streams to perform positional, range-based reads, which
+ * are more efficient than unbounded reads in many cloud provider object stores.
+ *
+ * Thread safety is not a requirement of the interface and is left to the
+ * implementation.
+ *
+ * If the implementation is also a {@link SeekableInputStream}, the position
+ * of the stream is not required to be updated based on the positional reads
+ * performed by this interface.  Usage of {@link SeekableInputStream} should
+ * always seek to the appropriate position for {@link java.io.InputStream}
+ * based reads.
+ *
+ */
+public interface RangeReadable extends Closeable {
+
+  /**
+   * Fill the provided buffer with the contents of the input source starting
+   * at {@code position} for the given {@code offset} and {@code length}.
+   *
+   * @param position start position of the read
+   * @param buffer target buffer to copy data
+   * @param offset offset in the buffer to copy the data
+   * @param length size of the read
+   */
+  void readFully(long position, byte[] buffer, int offset, int length) throws IOException;
```

Comment:
What will happen if the expected `length` exceed the object max file size starting from the given `position` ?  Will S3 SDK throw an IOException, or just read the remaining bytes silently ? 

If it's the later case, then I think we'd better to return the real read size in this method.

Reply:
Okay,  looks like we are throwing the exception: https://github.com/apache/iceberg/pull/4608/files#diff-5a63855b4791f694674e376ee7e6c3186c7a9783704699fbfe70954fa7ce6659R41-R43